crates/test-programs/src/bin/preview1_fd_filestat_set.rs

```diff
@@ -81,5 +73,11 @@ fn main() {
     };
 
     // Run the tests.
-    unsafe { test_fd_filestat_set(dir_fd) }
+    //
+    // The filestat_set functions should behave the same whether the file
+    // descriptor is open for read-only or read-write, because the underlying
+    // permissions of the file determine whether or not the filestat can be
+    // set or not, not than the open mode.
+    unsafe { test_fd_filestat_set(dir_fd, wasi::RIGHTS_FD_READ) }
```

Comment:
This doesn't seem right. If we open the file with just `RIGHTS_FD_READ`, we shuldn't be able to do `fd_filestat_set_size` on it. That would be analogous in POSIX to opening a file with `FD_READ` and calling `ftruncate`, which is defined to fail.

Reply:
That was what I thought too but I didnt chase it down in the ftruncate case. From the python test suite I found that futime is expected to succeed even when a file is read-only. So, I'll have to change this test, as well as set-len to check the open mode and reject size changes when not open for writing (fail with EINVAL, which I prefer here to EBADF) but permit set-times.

Reply:
This problem ended up being down to how the test was written. wasi::OFLAGS_CREAT implies that the WASI impl open the file with the O_RRONLY or O_RDWR in posix, so we were never actually opening a file just for reading in this test. I changed the test to create the file, then open it under the requested mode, and it behaves as expected.